lib/internal/modules/esm/loader.js

```diff
@@ -223,6 +223,25 @@ class ModuleLoader {
     };
   }
 
+  /**
+   * Evict saved result of `resolve` and `load` for the given import parameters.
+   */
+  evict(specifier, parentURL, importAttributes = {}) {
+    let resolved;
+    try {
+      resolved = this.resolveSync(specifier, parentURL, importAttributes);
+    } catch {
+      return false;
+    }
+    const requestKey = this.#resolveCache.serializeKey(specifier, importAttributes);
+    let didEvict = this.#resolveCache.delete(requestKey, parentURL);
+    if (this.loadCache.delete(resolved.url, importAttributes.type)) {
+      // nb: Careful with short-circuits here, we want to run both deletes unconditionally.
+      didEvict = true;
+    }
+    return didEvict;
```

Comment:
Should we perhaps evict from the load cache first? I know these calls are sync, but the load _hook_ is not; if `evict` is called while a module is being loaded (so after it’s resolved but before `load` has finished and it’s been added to the load cache) then a module could end up deleted from the resolve cache but not from the load cache. Whereas if you delete from the load cache first, and only on success of that do you continue on to delete from the resolve cache, we avoid leaving the caches in a “broken” state where the resolution stays cached by the load got uncached.

This raises a broader question: do we need to delete from the resolve cache? What are the use cases for this API, and could they be fulfilled by deleting only from the load cache? As in, when would we need to redo the resolution? Perhaps after new module customization hooks were registered, that would affect resolution?

Reply:
Deleting the resolution is nice because otherwise updates to package.json cannot be hot loaded and would force an app restart. But deleting the resolve cache in the manner in my diff is definitely a half measure. I think a more reliable feature would need to record all the resolutions to a given loaded module, so that when the loaded module is evicted *all* resolutions to it are deleted as well. The abstract specification method `HostLoadImportedModule` is decomposed into our `resolve` and `load` stages-- each with their own cache. To actually clear that abstract method requires more machinery which I'm not sure is worth it.